### PR TITLE
z_skin_matrix OK

### PR DESCRIPTION
**Pull request by Rozelette**
File is mostly copied from OOT with some new rotation functions at the end of the file. The only function that needed changes was `SkinMatrix_Invert`, which I added a `SWAP` macro for. It feels like MM is much more sensitive about scopes. All documentation is copied from OOT.

### CODE REVIEW

**Review by Thar0 — APPROVED** [review]
Nice work on this file.
It is sad to see they removed the completely unused quaternion function seen in OoT.

File: tables/variables.txt [hunk 1]
```diff
@@ -4430,7 +4430,7 @@
     0x801FBDE0:("D_801FBDE0","UNK_TYPE1","",0x1),
     0x801FBDE8:("D_801FBDE8","Vec3f","",0xc),
     0x801FBE00:("sMatrixStack","MtxF*","",0x4),
-    0x801FBE04:("sCurrentMatrix","z_Matrix*","",0x4),
+    0x801FBE04:("sCurrentMatrix","MtxF*","",0x4),
```
Comment (Rozelette) [inline tables/variables.txt:4433]: Test comment to test githook integrations

**Review by Rozelette — COMMENTED** [review]
Test review to test githook integrations

**Comment by Kenix3** [comment]
Looks good to me!

**Review by Kenix3 — APPROVED** [review]

File: src/code/z_skin_matrix.c [hunk 2]
```diff
@@ -0,0 +1,804 @@
...
+    mf->yx = 0.0f;
+    mf->yw = 0.0f;
+
+    mf->zx = 0.0f;
+    mf->zw = 0.0f;
+
+    mf->wx = 0.0f;
+    mf->wy = 0.0f;
+    mf->wz = 0.0f;
+
```
Comment (m4xw) [inline src/code/z_skin_matrix.c:659]: extra \n

File: src/code/z_skin_matrix.c [hunk 3]
```diff
@@ -0,0 +1,804 @@
...
+    s32 pad;
+    Mtx* mtx;
+
+    // TODO allocation should be a macro
+    mtx = (Mtx*)((int)gfxCtx->polyOpa.d - sizeof(Mtx));
+    gfxCtx->polyOpa.d = (void*)mtx;
+
+    if (mtx == NULL) {
+        return NULL;
+    }
```
Comment (m4xw) [inline src/code/z_skin_matrix.c:586]: add \n after }?

File: src/code/z_skin_matrix.c [hunk 4]
```diff
@@ -0,0 +1,804 @@
...
+        cos2 = Math_CosS(yaw);
+
+        mf->zz = cos * cos2;
+        mf->yz = cos * sin2;
+
+        mf->zx = sin * cos2;
+        mf->yx = sin * sin2;
+        mf->zy = -sin2;
+        mf->yy = cos2;
+
```
Comment (m4xw) [inline src/code/z_skin_matrix.c:399]: extra \n

File: src/code/z_skin_matrix.c [hunk 5]
```diff
@@ -0,0 +1,804 @@
...
+ */
+void SkinMatrix_SetRotateYRP(MtxF* mf, s16 yaw, s16 roll, s16 pitch) {
+    f32 cos2;
+    f32 sin;
+    f32 cos;
+    f32 xz;
+    f32 sin2;
+    f32 yz;
+    f32 xx;
+    f32 yx;
```
Comment (m4xw) [inline src/code/z_skin_matrix.c:375]: \n after locals

File: src/code/z_skin_matrix.c [hunk 6]
```diff
@@ -0,0 +1,804 @@
...
+    sin = Math_SinS(roll);
+    cos = Math_CosS(roll);
```
Comment (m4xw) [inline src/code/z_skin_matrix.c:377]: \n before mf setters

File: src/code/z_skin_matrix.c [hunk 7]
```diff
@@ -0,0 +1,804 @@
...
+        cos2 = Math_CosS(pitch);
+
+        mf->xx = cos * cos2;
+        mf->zx = cos * sin2;
+
+        mf->xy = sin * cos2;
+        mf->zy = sin * sin2;
+        mf->xz = -sin2;
+        mf->zz = cos2;
+
```
Comment (m4xw) [inline src/code/z_skin_matrix.c:330]: extra \n

File: src/code/z_skin_matrix.c [hunk 8]
```diff
@@ -0,0 +1,804 @@
...
+        for (thisCol = 0; thisCol < 4; thisCol++) {
+            if (thisCol != thisRow) {
+                temp2 = mfCopy.mf[thisRow][thisCol];
+                for (i = 0; i < 4; i++) {
+                    mfCopy.mf[i][thisCol] -= mfCopy.mf[i][thisRow] * temp2;
+                    dest->mf[i][thisCol] -= dest->mf[i][thisRow] * temp2;
+                }
+            }
+        }
+    }
```
Comment (m4xw) [inline src/code/z_skin_matrix.c:274]: add \n

File: src/code/z_skin_matrix.c [hunk 9]
```diff
@@ -0,0 +1,804 @@
...
+ */
+s32 SkinMatrix_Invert(MtxF* src, MtxF* dest) {
+    MtxF mfCopy;
+    s32 i;
+    f32 temp2;
+    s32 thisRow;
+    s32 thisCol;
+
+    SkinMatrix_MtxFCopy(src, &mfCopy);
+    SkinMatrix_Clear(dest);
```
Comment (m4xw) [inline src/code/z_skin_matrix.c:240]: \n after this line

File: src/code/z_skin_matrix.c [hunk 10]
```diff
@@ -0,0 +1,804 @@
...
+
+/**
+ * Matrix multiplication, dest = mfA * mfB.
+ * mfA and dest should not be the same matrix.
+ */
+void SkinMatrix_MtxFMtxFMult(MtxF* mfB, MtxF* mfA, MtxF* dest) {
+    f32 rx;
+    f32 ry;
+    f32 rz;
+    f32 rw;
```
Comment (m4xw) [inline src/code/z_skin_matrix.c:58]: add \n after locals

File: src/code/z_skin_matrix.c [hunk 11]
```diff
@@ -0,0 +1,804 @@
...
+ *
+ * \f[ [\texttt{dest}, -] = [\texttt{src}, 1] \cdot [mf] \f]
+ */
+void SkinMatrix_Vec3fMtxFMultXYZ(MtxF* mf, Vec3f* src, Vec3f* dest) {
+    f32 mx = mf->xx;
+    f32 my = mf->yx;
+    f32 mz = mf->zx;
+    f32 mw = mf->wx;
+
+    dest->x = mw + ((src->x * mx) + (src->y * my) + (src->z * mz));
```
Comment (m4xw) [inline src/code/z_skin_matrix.c:37]: I suggest putting the dest->x without the extra \n next to the local setter, add a \n after dest->

File: src/code/z_skin_matrix.c [hunk 12]
```diff
@@ -0,0 +1,804 @@
...
+    mx = mf->xy;
+    my = mf->yy;
+    mz = mf->zy;
+    mw = mf->wy;
+    dest->y = mw + ((src->x * mx) + (src->y * my) + (src->z * mz));
```
Comment (m4xw) [inline src/code/z_skin_matrix.c:42]: \n after this

File: src/code/z_skin_matrix.c [hunk 13]
```diff
@@ -0,0 +1,804 @@
...
+    mf->yz = 0.0f;
+    mf->yw = 0.0f;
+
+    mf->zy = 0.0f;
+    mf->zw = 0.0f;
+
+    mf->wx = 0.0f;
+    mf->wy = 0.0f;
+    mf->wz = 0.0f;
+
```
Comment (m4xw) [inline src/code/z_skin_matrix.c:727]: extra \n

**Review by m4xw — COMMENTED** [review]
nits

**Pull request merged**